// testcafe/lib/test-run/role-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProxyRoleProvider = exports.ProxylessRoleProvider = void 0;
const testcafe_hammerhead_1 = require("testcafe-hammerhead");
const cookie_provider_1 = require("../proxyless/cookie-provider");
class RoleProviderBase {
    constructor(testRun) {
        this._testRun = testRun;
    }
    get session() {
        return this._testRun.session;
    }
}
class ProxylessRoleProvider extends RoleProviderBase {
    constructor(testRun) {
        super(testRun);
        this._cookieProvider = new cookie_provider_1.CdpCookieProvider(testRun);
    }
    async getStateSnapshot() {
        const cookies = await this._cookieProvider.getCookies([]);
        return new testcafe_hammerhead_1.StateSnapshot(JSON.stringify(cookies), null);
    }
    async useStateSnapshot(stateSnapshot) {
        const { cookies } = stateSnapshot;
        await this._cookieProvider.deleteCookies();
        if (cookies)
            await this._cookieProvider.setCookies(JSON.parse(cookies), '');
        this._testRun.saveStoragesSnapshot(stateSnapshot.storages);
    }
}
exports.ProxylessRoleProvider = ProxylessRoleProvider;
class ProxyRoleProvider extends RoleProviderBase {
    async getStateSnapshot() {
        return this.session.getStateSnapshot();
    }
    async useStateSnapshot(stateSnapshot) {
        await this.session.useStateSnapshot(stateSnapshot);
    }
}
exports.ProxyRoleProvider = ProxyRoleProvider;
