// testcafe/lib/utils/get-options/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSkipJsErrorsOptions = exports.getDashboardOptions = exports.getCompilerOptions = exports.getGrepOptions = exports.getMetaOptions = exports.getVideoOptions = exports.getScreenshotOptions = exports.getQuarantineOptions = exports.getSSLOptions = void 0;
const ssl_1 = __importDefault(require("./ssl"));
exports.getSSLOptions = ssl_1.default;
const quarantine_1 = require("./quarantine");
Object.defineProperty(exports, "getQuarantineOptions", { enumerable: true, get: function () { return quarantine_1.getQuarantineOptions; } });
const screenshot_1 = __importDefault(require("./screenshot"));
exports.getScreenshotOptions = screenshot_1.default;
const video_1 = __importDefault(require("./video"));
exports.getVideoOptions = video_1.default;
const meta_1 = __importDefault(require("./meta"));
exports.getMetaOptions = meta_1.default;
const grep_1 = __importDefault(require("./grep"));
exports.getGrepOptions = grep_1.default;
const compiler_1 = __importDefault(require("./compiler"));
exports.getCompilerOptions = compiler_1.default;
const dashboard_1 = __importDefault(require("./dashboard"));
exports.getDashboardOptions = dashboard_1.default;
const skip_js_errors_1 = require("./skip-js-errors");
Object.defineProperty(exports, "getSkipJsErrorsOptions", { enumerable: true, get: function () { return skip_js_errors_1.getSkipJsErrorsOptions; } });
