// testcafe/lib/proxyless/utils/cdp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRequestId = exports.createRequestPausedEventForResponse = exports.isPreflightRequest = exports.isRequestPausedEvent = exports.isUnauthorized = exports.isRequest = exports.dispatchEvent = exports.navigateTo = exports.redirect = void 0;
const http_status_codes_1 = require("http-status-codes");
const headers_1 = require("./headers");
const types_1 = require("../types");
const errors_1 = require("../errors");
async function redirect(client, requestId, url) {
    await client.Fetch.fulfillRequest({
        requestId,
        responseCode: http_status_codes_1.StatusCodes.MOVED_PERMANENTLY,
        responseHeaders: [
            { name: 'location', value: url },
        ],
    });
}
exports.redirect = redirect;
async function navigateTo(client, url) {
    await client.Page.navigate({ url });
}
exports.navigateTo = navigateTo;
async function dispatchEvent(client, type, options) {
    switch (+type) {
        case types_1.EventType.Mouse:
            await client.Input.dispatchMouseEvent(options);
            break;
        case types_1.EventType.Keyboard:
            await client.Input.dispatchKeyEvent(options);
            break;
        case types_1.EventType.Touch:
            await client.Input.dispatchTouchEvent(options);
            break;
        default:
            throw (0, errors_1.unknownCDPEventType)(options.type);
    }
}
exports.dispatchEvent = dispatchEvent;
function isRequest(event) {
    return event.responseStatusCode === void 0;
}
exports.isRequest = isRequest;
function isUnauthorized(statusCode) {
    return statusCode === 401;
}
exports.isUnauthorized = isUnauthorized;
function isRequestPausedEvent(val) {
    return val && val.frameId && typeof val.request === 'object';
}
exports.isRequestPausedEvent = isRequestPausedEvent;
function isPreflightRequest(event) {
    return event.request.method === 'OPTIONS';
}
exports.isPreflightRequest = isPreflightRequest;
function createRequestPausedEventForResponse(mockedResponse, requestEvent) {
    return Object.assign({}, requestEvent, {
        responseStatusCode: mockedResponse.statusCode,
        responseHeaders: (0, headers_1.convertToHeaderEntries)(mockedResponse.headers),
    });
}
exports.createRequestPausedEventForResponse = createRequestPausedEventForResponse;
function getRequestId(event) {
    if (isRequestPausedEvent(event))
        return event.networkId;
    return event.frame.loaderId;
}
exports.getRequestId = getRequestId;
