// testcafe/lib/client/driver/status.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assignable_1 = __importDefault(require("../../utils/assignable"));
const generate_id_1 = __importDefault(require("./generate-id"));
class DriverStatus extends assignable_1.default {
    constructor(obj) {
        super(obj);
        this.id = (0, generate_id_1.default)();
        this.isCommandResult = false;
        this.executionError = null;
        this.pageError = null;
        this.resent = false;
        this.result = null;
        this.consoleMessages = null;
        this.isPendingWindowSwitching = false;
        this.isObservingFileDownloadingInNewWindow = false;
        this.isFirstRequestAfterWindowSwitching = false;
        this.debug = '';
        this.warnings = null;
        this._assignFrom(obj, true);
    }
    getAssignableProperties() {
        return [
            { name: 'isCommandResult' },
            { name: 'executionError' },
            { name: 'pageError' },
            { name: 'result' },
            { name: 'consoleMessages' },
            { name: 'isPendingWindowSwitching' },
            { name: 'isObservingFileDownloadingInNewWindow' },
            { name: 'isFirstRequestAfterWindowSwitching' },
            { name: 'warnings' },
        ];
    }
}
exports.default = DriverStatus;
