// testcafe/lib/services/utils/ipc/interfaces.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isIPCResponsePacket = exports.isIPCErrorResponse = exports.isTestCafeErrorList = exports.IPCTransportEvents = exports.IPCPacketType = void 0;
const error_list_1 = __importDefault(require("../../../errors/error-list"));
var IPCPacketType;
(function (IPCPacketType) {
    IPCPacketType[IPCPacketType["request"] = 0] = "request";
    IPCPacketType[IPCPacketType["response"] = 1] = "response";
})(IPCPacketType = exports.IPCPacketType || (exports.IPCPacketType = {}));
var IPCTransportEvents;
(function (IPCTransportEvents) {
    IPCTransportEvents["data"] = "data";
})(IPCTransportEvents = exports.IPCTransportEvents || (exports.IPCTransportEvents = {}));
function isTestCafeErrorList(err) {
    return err.name === error_list_1.default.name;
}
exports.isTestCafeErrorList = isTestCafeErrorList;
function isIPCErrorResponse(response) {
    return !!response.error;
}
exports.isIPCErrorResponse = isIPCErrorResponse;
function isIPCResponsePacket(message) {
    return ['id', 'type', 'sync', 'data'].every(prop => prop in message);
}
exports.isIPCResponsePacket = isIPCResponsePacket;
