// testcafe/lib/services/utils/ipc/transport.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceTransport = exports.HostTransport = void 0;
const io_1 = require("./io");
const async_event_emitter_1 = __importDefault(require("../../../utils/async-event-emitter"));
const runtime_1 = require("../../../errors/runtime");
const types_1 = require("../../../errors/types");
const interfaces_1 = require("./interfaces");
class HostTransport extends async_event_emitter_1.default {
    constructor(inputStream, outputStream, syncStream) {
        super();
        this.asyncReader = new io_1.AsyncReader(inputStream);
        this.asyncWriter = new io_1.AsyncWriter(outputStream);
        this.syncReader = new io_1.AsyncReader(syncStream);
        this.syncWriter = new io_1.AsyncWriter(syncStream);
        this.readers = [this.asyncReader, this.syncReader];
    }
    read() {
        this.readers.forEach(reader => {
            reader.on('data', data => this.emit('data', data));
            reader.read();
        });
    }
    async write(message) {
        const writer = message.sync ? this.syncWriter : this.asyncWriter;
        await writer.write(message);
    }
    readSync() {
        throw new runtime_1.GeneralError(types_1.RUNTIME_ERRORS.methodIsNotAvailableForAnIPCHost);
    }
    writeSync() {
        throw new runtime_1.GeneralError(types_1.RUNTIME_ERRORS.methodIsNotAvailableForAnIPCHost);
    }
}
exports.HostTransport = HostTransport;
class ServiceTransport extends async_event_emitter_1.default {
    constructor(inputStream, outputStream, syncFd) {
        super();
        this.asyncReader = new io_1.AsyncReader(inputStream);
        this.asyncWriter = new io_1.AsyncWriter(outputStream);
        this.syncReader = new io_1.SyncReader(syncFd);
        this.syncWriter = new io_1.SyncWriter(syncFd);
    }
    read() {
        this.asyncReader.on('data', data => this.emit('data', data));
        this.asyncReader.read();
    }
    async write(message) {
        await this.asyncWriter.write(message);
    }
    readSync() {
        const message = this.syncReader.readSync();
        if (!(0, interfaces_1.isIPCResponsePacket)(message))
            throw new runtime_1.GeneralError(types_1.RUNTIME_ERRORS.malformedIPCMessage);
        return message;
    }
    writeSync(message) {
        this.syncWriter.writeSync(message);
    }
}
exports.ServiceTransport = ServiceTransport;
