// testcafe/lib/compiler/test-file/formats/dev-tools/compiler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const compiler_1 = __importDefault(require("../raw/compiler"));
const factory_1 = require("./commands/factory");
const switch_to_iframe_1 = require("./commands/switch-to-iframe");
const switch_to_main_window_1 = require("./commands/switch-to-main-window");
const TEST_BASE = JSON.stringify({
    fixtures: [
        {
            name: 'New Fixture',
            tests: [
                {
                    name: 'New Test',
                    commands: [],
                },
            ],
        },
    ],
});
class DevToolsTestFileCompiler extends compiler_1.default {
    constructor() {
        super(...arguments);
        this.raw = { fixtures: [] };
    }
    _hasTests() {
        return true;
    }
    get _fixture() {
        return this.raw.fixtures[0];
    }
    get _test() {
        return this._fixture.tests[0];
    }
    getSupportedExtension() {
        return '.json';
    }
    compile(code, filename) {
        this.raw = JSON.parse(TEST_BASE);
        const preprocessedCode = this._preProcess(code, filename);
        if (!preprocessedCode)
            return [];
        return super.compile(preprocessedCode, filename);
    }
    _preProcess(code, filename) {
        const parsedCode = JSON.parse(code);
        this._fixture.name = parsedCode.title;
        this._test.name = parsedCode.title;
        if (!parsedCode.steps)
            return null;
        parsedCode.steps.forEach((step, i) => this._processStep(step, filename, i));
        return JSON.stringify(this.raw);
    }
    _processStep(step, filename, i) {
        const transformer = factory_1.CommandTransformerFactory.create(step, filename, i);
        if (transformer) {
            this._onBeforeCommandExecute(step);
            this._test.commands.push(transformer.transform());
            this._onAfterCommandExecute(step);
        }
    }
    _onBeforeCommandExecute(step) {
        if (!step.frame)
            return;
        const frames = step.frame;
        for (const frame of frames)
            this._test.commands.push(new switch_to_iframe_1.SwitchToIframeCommandTransformer({ frame }, 0).transform());
    }
    _onAfterCommandExecute(step) {
        this._test.commands.push(new switch_to_main_window_1.SwitchToMainWindowCommandTransformer(step, 0).transform());
    }
}
exports.default = DevToolsTestFileCompiler;
module.exports = exports.default;
