// testcafe-hammerhead/lib/request-pipeline/request-hooks/events/request-event.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const generate_unique_id_1 = __importDefault(require("../../../utils/generate-unique-id"));
class RequestEvent {
    constructor(init) {
        Object.assign(this, init);
        this.id = (0, generate_unique_id_1.default)();
    }
    async setMock(mock) {
        await this.setMockFn(this.id, mock);
    }
    get requestOptions() {
        return this.reqOpts;
    }
    get isAjax() {
        return this._requestInfo.isAjax;
    }
    static from(data) {
        const { id, requestFilterRule, _requestInfo } = data;
        const requestEvent = new RequestEvent({
            requestFilterRule: requestFilterRule,
            reqOpts: {},
            setMockFn: () => Promise.resolve(),
            _requestInfo,
        });
        requestEvent.id = id;
        return requestEvent;
    }
}
exports.default = RequestEvent;module.exports = exports.default;
