// testcafe/lib/proxyless/request-hooks/event-factory/frame-navigated-event-based.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const testcafe_hammerhead_1 = require("testcafe-hammerhead");
const cdp_1 = require("../../utils/cdp");
class FrameNavigatedEventBasedEventFactory extends testcafe_hammerhead_1.BaseRequestHookEventFactory {
    constructor(event, sessionId) {
        super();
        this._event = event;
        this._sessionId = sessionId;
    }
    createRequestInfo() {
        // NOTE: We can't get some request information from the FrameNavigated event.
        // So, we initialize the RequestInfo object with empty or default values.
        return new testcafe_hammerhead_1.RequestInfo({
            requestId: (0, cdp_1.getRequestId)(this._event),
            sessionId: this._sessionId,
            userAgent: '',
            url: this._event.frame.url,
            method: '',
            headers: {},
            body: Buffer.alloc(0),
            isAjax: false,
        });
    }
    createConfigureResponseEvent(rule) {
        // NOTE: Used as a stub
        return new testcafe_hammerhead_1.ConfigureResponseEvent(rule, {});
    }
    createRequestOptions() {
        // NOTE: Used as a stub
        return {};
    }
    createResponseInfo() {
        // NOTE: Used as a stub
        return {};
    }
}
exports.default = FrameNavigatedEventBasedEventFactory;
module.exports = exports.default;
