// testcafe/lib/proxyless/client/input.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../types");
const axis_values_1 = __importDefault(require("../../client/core/utils/values/axis-values"));
const event_descriptor_1 = __importDefault(require("./event-descriptor"));
class ProxylessInput {
    constructor(dispatchEventFn, leftTopPoint) {
        this._dispatchEventFn = dispatchEventFn;
        this._leftTopPoint = leftTopPoint || new axis_values_1.default(0, 0);
    }
    mouseDown(options) {
        const eventOptions = event_descriptor_1.default.createMouseEventOptions('mousePressed', options, this._leftTopPoint);
        return this._dispatchEventFn.single(types_1.EventType.Mouse, eventOptions);
    }
    mouseUp(options) {
        const eventOptions = event_descriptor_1.default.createMouseEventOptions('mouseReleased', options, this._leftTopPoint);
        return this._dispatchEventFn.single(types_1.EventType.Mouse, eventOptions);
    }
    keyDown(options) {
        const eventOptions = event_descriptor_1.default.createKeyDownOptions(options);
        return this._dispatchEventFn.single(types_1.EventType.Keyboard, eventOptions);
    }
    keyUp(options) {
        const eventOptions = event_descriptor_1.default.createKeyUpOptions(options);
        return this._dispatchEventFn.single(types_1.EventType.Keyboard, eventOptions);
    }
    executeEventSequence(eventSequence) {
        return this._dispatchEventFn.sequence(eventSequence);
    }
}
exports.default = ProxylessInput;
module.exports = exports.default;
